crates/signer/src/signer.rs

```diff
@@ -13,17 +13,56 @@ use alloy_sol_types::{Eip712Domain, SolStruct};
 /// receipt type, and is blanket implemented for transactions which our signer
 /// trait can produce signatures for.
 pub trait SignableTx: Send + Sync + 'static {
+    /// Encode the transaction.
+    fn rlp_encode(&self) -> Vec<u8>;
+
     /// Calculate the signing hash for the transaction.
     fn signature_hash(&self) -> B256;
+
+    /// Returns the chain ID. Used for EIP-155 signing.
+    fn chain_id(&self) -> Option<ChainId>;
+
+    /// Sets the chain ID. Used for EIP-155 signing.
+    fn set_chain_id(&mut self, chain_id: ChainId);
+
+    /// Set `chain_id` if it is not already set. Checks that the provided `chain_id` matches the
+    /// existing `chain_id` if it is already set.
+    fn set_chain_id_checked(&mut self, chain_id: ChainId) -> Result<()> {
+        match self.chain_id() {
+            Some(tx_chain_id) => {
+                if tx_chain_id != chain_id {
+                    return Err(crate::Error::TransactionChainIdMismatch {
+                        signer: chain_id,
+                        tx: tx_chain_id,
+                    });
+                }
+            }
+            None => {
+                self.set_chain_id(chain_id);
+            }
+        }
```

Comment:
can simplify with

```rust
let Some(tx_chain_id)  = self.chain_id()  { else self.set_chain_Id(chain_id); return Ok(()) };
```

and pull the error check outside

Reply:
don't think it makes it any simple